# REDES NEURONALES

In [2]:
#Importando librerias

import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.model_selection import train_test_split

Se genero un excel con los datos consolidados que se utilizaron en R para importarlos en este ambiente.

In [3]:
datos_consolidados = pd.read_excel('datos_consolidados.xlsx')

In [4]:
datos_consolidados.head()

núm_corre  año_ocu  mes_ocu  día_ocu  día_sem_ocu  hora_ocu  g_hora  \
0          1     2015        1        1            4        16       3   
1          2     2015        1        1            4        16       3   
2          3     2015        1        1            4        22       4   
3          4     2015        1        1            4         2       1   
4          5     2015        1        1            4         9       2   

   g_hora_5  depto_ocu  mupio_ocu  ...  edad_quinquenales  fall_les  \
0         2          1        101  ...                  4         2   
1         2          1        101  ...                  4         2   
2         3          1        101  ...                  7         2   
3         1          1        101  ...                  4         2   
4         1          1        101  ...                  5         1   

   int_o_noint  tipo_veh  marca_veh  color_veh  modelo_veh  g_modelo_veh  \
0            1         4         21          5        2011             5   
1            1         4         21          5        2011             5   
2            1         4         21          5        9999             6   
3            1         4         21          1        2000             4   
4            2         4         40          5        9999             6   

   tipo_eve  horario  
0         1        1  
1         1        1  
2         1        2  
3         1        2  
4         2        1  

[5 rows x 26 columns]

In [13]:
datos_consolidados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91367 entries, 0 to 91366
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   núm_corre          91367 non-null  int64
 1   año_ocu            91367 non-null  int64
 2   mes_ocu            91367 non-null  int64
 3   día_ocu            91367 non-null  int64
 4   día_sem_ocu        91367 non-null  int64
 5   hora_ocu           91367 non-null  int64
 6   g_hora             91367 non-null  int64
 7   g_hora_5           91367 non-null  int64
 8   depto_ocu          91367 non-null  int64
 9   mupio_ocu          91367 non-null  int64
 10  zona_ocu           91367 non-null  int64
 11  sexo_per           91367 non-null  int64
 12  edad_per           91367 non-null  int64
 13  mayor_menor        91367 non-null  int64
 14  g_edad_80ymás      91367 non-null  int64
 15  g_edad_60ymás      91367 non-null  int64
 16  edad_quinquenales  91367 non-null  int64
 17  fall_les    

In [14]:
datos_consolidados.describe()

núm_corre       año_ocu       mes_ocu       día_ocu   día_sem_ocu  \
count  91367.000000  91367.000000  91367.000000  91367.000000  91367.000000   
mean    5137.297700   2019.032167      6.527072     15.722274      4.356245   
std     3029.321818      2.649056      3.537576      8.862713      2.064672   
min        1.000000   2015.000000      1.000000      1.000000      1.000000   
25%     2538.500000   2017.000000      3.000000      8.000000      3.000000   
50%     5076.000000   2019.000000      7.000000     16.000000      5.000000   
75%     7614.000000   2021.000000     10.000000     23.000000      6.000000   
max    11668.000000   2023.000000     12.000000     31.000000      7.000000   

           hora_ocu        g_hora      g_hora_5     depto_ocu     mupio_ocu  \
count  91367.000000  91367.000000  91367.000000  91367.000000  91367.000000   
mean      13.554215      2.847264      1.993838      7.744131    778.921416   
std        6.849150      1.063200      0.844961      6.961212    695.779349   
min        0.000000      1.000000      1.000000      1.000000    101.000000   
25%        8.000000      2.000000      1.000000      1.000000    109.000000   
50%       15.000000      3.000000      2.000000      5.000000    506.000000   
75%       19.000000      4.000000      3.000000     14.000000   1413.000000   
max       99.000000      5.000000      4.000000     22.000000   2217.000000   

       ...  edad_quinquenales      fall_les   int_o_noint      tipo_veh  \
count  ...       91367.000000  91367.000000  91367.000000  91367.000000   
mean   ...           7.693949      1.819278      1.263235      9.528243   
std    ...           4.332270      0.388047      0.614172     21.587794   
min    ...           1.000000      1.000000      1.000000      1.000000   
25%    ...           5.000000      2.000000      1.000000      3.000000   
50%    ...           6.000000      2.000000      1.000000      4.000000   
75%    ...           9.000000      2.000000      1.000000      4.000000   
max    ...          18.000000      9.000000      9.000000     99.000000   

          marca_veh     color_veh    modelo_veh  g_modelo_veh      tipo_eve  \
count  91367.000000  91367.000000  91367.000000  91367.000000  91367.000000   
mean     303.501220     29.492267   7641.748706     48.684208      2.647477   
std      424.957515     42.279423   3644.142801     46.837587      2.540236   
min        1.000000      1.000000   1898.000000      1.000000      1.000000   
25%       27.000000      2.000000   2019.000000      5.000000      1.000000   
50%       67.000000      5.000000   9999.000000      6.000000      1.000000   
75%      999.000000     99.000000   9999.000000     99.000000      5.000000   
max      999.000000     99.000000   9999.000000     99.000000     99.000000   

            horario  
count  91367.000000  
mean       1.435102  
std        0.495773  
min        1.000000  
25%        1.000000  
50%        1.000000  
75%        2.000000  
max        2.000000  

[8 rows x 26 columns]

## Primer modelo: Predecir tipo de evento

Se realizan los mismos filtros y se trabaja con un modelo de prediccion no binario para comparar los resultados obtenidos con otros modelos.

In [74]:
data_arbol2 = datos_consolidados[
    (datos_consolidados["fall_les"].isin([1, 2])) &
    (datos_consolidados["sexo_per"] != 9) &
    (datos_consolidados["tipo_eve"] != 99) &
    (datos_consolidados["tipo_eve"] != 1) &
       (datos_consolidados["tipo_veh"] != 99) &
    (datos_consolidados["edad_per"] != 999)].copy()


In [77]:
X = data_arbol2[['sexo_per','horario', 'mayor_menor','tipo_veh','depto_ocu','fall_les']]
X = X.fillna(0)
y = data_arbol2['tipo_eve'].values - 1

In [78]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [79]:

model = Sequential()
model.add(Dense(8, input_dim=6, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(8, activation='softmax'))

C:\Users\Brandon Portillo\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Esta red contiene 6 variables de entrada y 8 salidas debido a que se trata de 8 casos a analizar.

In [80]:

model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [81]:

model.fit(X_train, y_train, epochs=50, batch_size=64, validation_data=(X_test, y_test))

Epoch 1/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 8s 11ms/step - accuracy: 0.4167 - loss: 1.5730 - val_accuracy: 0.4405 - val_loss: 1.4515
Epoch 2/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.4430 - loss: 1.4295 - val_accuracy: 0.4479 - val_loss: 1.4173
Epoch 3/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.4481 - loss: 1.4061 - val_accuracy: 0.4511 - val_loss: 1.4067
Epoch 4/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.4551 - loss: 1.3906 - val_accuracy: 0.4577 - val_loss: 1.3875
Epoch 5/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.4605 - loss: 1.3759 - val_accuracy: 0.4586 - val_loss: 1.3747
Epoch 6/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.4619 - loss: 1.3633 - val_accuracy: 0.4562 - val_loss: 1.3674
Epoch 7/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.4622 - loss: 1.3569 - val_accuracy: 0.4575 - val_loss: 1.3617
Epoch 8/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.4641 - loss: 1.3513 - val_accuracy: 0

In [82]:

loss, accuracy = model.evaluate(X_test, y_test)
print(f'Loss: {loss}, Accuracy: {accuracy}')

238/238 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.4853 - loss: 1.2966
Loss: 1.2965670824050903, Accuracy: 0.4853154122829437


El modelo no acierta el 46.8% de las veces

In [83]:
caso1= np.array([[2,1,1,4,1,1]])
caso2= np.array([[1,2,2,1,6,1]])
caso3= np.array([[2,1,2,3,14,2]])
caso4= np.array([[1,1,1,2,18,2]])
caso5= np.array([[2,2,1,5,22,2]])
casos=[caso1,caso2,caso3,caso4,caso5]

In [84]:
predicciones_arbol2= []
for i in casos:
    predicciones_arbol2.append(model.predict(i))

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 203ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 117ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 109ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 112ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 105ms/step


In [85]:
pred_array = np.vstack(predicciones_arbol2)   # queda shape (n, 22)

df_prediccion = pd.DataFrame(
    np.round(pred_array, 3),                 # Redondear a 3 decimales
    index = np.arange(1, pred_array.shape[0] + 1),   # Filas desde 1
    columns = np.arange(1, pred_array.shape[1] + 1)  # Columnas desde 1
)

df_prediccion

1      2      3      4      5      6      7      8
1  0.0  0.156  0.025  0.040  0.511  0.265  0.002  0.000
2  0.0  0.171  0.079  0.009  0.725  0.001  0.015  0.001
3  0.0  0.174  0.312  0.066  0.377  0.008  0.062  0.001
4  0.0  0.220  0.261  0.053  0.409  0.008  0.048  0.000
5  0.0  0.267  0.148  0.125  0.341  0.099  0.021  0.000

## Modelo 2: Prediccion de tipo de vehiculo

In [64]:
data_arbol3 = datos_consolidados[
    (datos_consolidados["fall_les"].isin([1, 2])) &
    (datos_consolidados["sexo_per"] != 9) &
    (datos_consolidados["tipo_eve"] != 99) &
    (datos_consolidados["tipo_veh"] != 4) &
       (datos_consolidados["tipo_veh"] != 99) &
    (datos_consolidados["edad_per"] != 999)].copy()

In [65]:
X = data_arbol3[['sexo_per','horario', 'mayor_menor','tipo_eve','depto_ocu','fall_les']]
X = X.fillna(0)
y = data_arbol3['tipo_veh'].values - 1

In [66]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Se tienen 24 casos de salida y 6 de entrada

In [67]:
model = Sequential()
model.add(Dense(16, input_dim=6, activation='relu'))
model.add(Dense(16, activation='relu'))
model.add(Dense(24, activation='softmax'))

C:\Users\Brandon Portillo\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [68]:

model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [69]:

model.fit(X_train, y_train, epochs=50, batch_size=64, validation_data=(X_test, y_test))

Epoch 1/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 6s 7ms/step - accuracy: 0.3131 - loss: 2.1965 - val_accuracy: 0.3535 - val_loss: 1.9607
Epoch 2/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.3569 - loss: 1.9436 - val_accuracy: 0.3620 - val_loss: 1.9298
Epoch 3/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.3592 - loss: 1.9277 - val_accuracy: 0.3605 - val_loss: 1.9199
Epoch 4/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.3618 - loss: 1.9200 - val_accuracy: 0.3610 - val_loss: 1.9159
Epoch 5/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.3619 - loss: 1.9154 - val_accuracy: 0.3631 - val_loss: 1.9148
Epoch 6/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.3634 - loss: 1.9105 - val_accuracy: 0.3640 - val_loss: 1.9081
Epoch 7/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.3658 - loss: 1.9055 - val_accuracy: 0.3643 - val_loss: 1.9048
Epoch 8/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.3658 - loss: 1.9016 - val_accuracy: 0.

In [70]:
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Loss: {loss}, Accuracy: {accuracy}')

254/254 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.3774 - loss: 1.8471
Loss: 1.8471425771713257, Accuracy: 0.37740740180015564


El modelo acerto el 37% de las veces

In [71]:
caso1= np.array([[2,1,1,2,1,1]])
caso2= np.array([[1,2,2,5,6,1]])
caso3= np.array([[2,1,2,3,14,2]])
caso4= np.array([[1,1,1,1,18,2]])
caso5= np.array([[2,2,1,6,22,2]])
casos=[caso1,caso2,caso3,caso4,caso5]

In [72]:
predicciones_arbol3= []
for i in casos:
    predicciones_arbol3.append(model.predict(i))


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 218ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 128ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 121ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 132ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 126ms/step


In [73]:
pred_array = np.vstack(predicciones_arbol3)   # queda shape (n, 22)

df_prediccion = pd.DataFrame(
    np.round(pred_array, 3),                 # Redondear a 3 decimales
    index = np.arange(1, pred_array.shape[0] + 1),   # Filas desde 1
    columns = np.arange(1, pred_array.shape[1] + 1)  # Columnas desde 1
)

df_prediccion

1      2      3    4      5      6      7      8      9      10  ...  \
1  0.304  0.095  0.188  0.0  0.046  0.011  0.037  0.001  0.045  0.003  ...   
2  0.467  0.126  0.190  0.0  0.081  0.048  0.008  0.003  0.009  0.002  ...   
3  0.026  0.057  0.518  0.0  0.048  0.002  0.027  0.000  0.094  0.000  ...   
4  0.155  0.105  0.302  0.0  0.070  0.023  0.017  0.006  0.167  0.001  ...   
5  0.356  0.095  0.320  0.0  0.080  0.024  0.008  0.002  0.056  0.001  ...   

    15   16     17     18   19   20     21     22   23   24  
1  0.0  0.0  0.001  0.003  0.0  0.0  0.000  0.005  0.0  0.0  
2  0.0  0.0  0.000  0.006  0.0  0.0  0.000  0.003  0.0  0.0  
3  0.0  0.0  0.000  0.000  0.0  0.0  0.000  0.009  0.0  0.0  
4  0.0  0.0  0.000  0.014  0.0  0.0  0.001  0.009  0.0  0.0  
5  0.0  0.0  0.000  0.003  0.0  0.0  0.000  0.004  0.0  0.0  

[5 rows x 24 columns]

## Modelo 3: Predecir horario (Dia/Noche)

In [53]:
data_arbol4 = datos_consolidados[
    (datos_consolidados["fall_les"].isin([1, 2])) &
    (datos_consolidados["sexo_per"] != 9) &
    (datos_consolidados["tipo_eve"] != 99) &
       (datos_consolidados["tipo_veh"] != 99) &
    (datos_consolidados["edad_per"] != 999)].copy()

In [54]:
X = data_arbol4[['sexo_per', 'mayor_menor','tipo_eve','tipo_veh','depto_ocu','fall_les']]
X = X.fillna(0)
y = data_arbol4['horario'].values - 1

In [55]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

En este analisis se tienen 2 casos de salida y 6 de entrada

In [56]:
model = Sequential()
model.add(Dense(8, input_dim=6, activation='relu'))
model.add(Dense(4, activation='relu'))
model.add(Dense(2, activation='softmax'))

C:\Users\Brandon Portillo\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [57]:

model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [58]:

model.fit(X_train, y_train, epochs=50, batch_size=64, validation_data=(X_test, y_test))

Epoch 1/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 8s 6ms/step - accuracy: 0.5722 - loss: 0.6786 - val_accuracy: 0.5783 - val_loss: 0.6745
Epoch 2/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.5837 - loss: 0.6724 - val_accuracy: 0.5797 - val_loss: 0.6741
Epoch 3/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.5859 - loss: 0.6710 - val_accuracy: 0.5864 - val_loss: 0.6722
Epoch 4/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.5891 - loss: 0.6702 - val_accuracy: 0.5892 - val_loss: 0.6714
Epoch 5/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.5898 - loss: 0.6692 - val_accuracy: 0.5837 - val_loss: 0.6709
Epoch 6/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.5918 - loss: 0.6681 - val_accuracy: 0.5902 - val_loss: 0.6690
Epoch 7/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.5930 - loss: 0.6673 - val_accuracy: 0.5878 - val_loss: 0.6685
Epoch 8/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.5935 - loss: 0.6669 - val_accuracy: 0.

In [60]:
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Loss: {loss}, Accuracy: {accuracy}')

499/499 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.5968 - loss: 0.6661
Loss: 0.6661440134048462, Accuracy: 0.5967630743980408


El modelo acerto el 59% de las veces

In [61]:
caso1= np.array([[2,1,2,4,1,1]])
caso2= np.array([[1,2,5,1,6,1]])
caso3= np.array([[2,2,3,3,14,2]])
caso4= np.array([[1,1,1,2,18,2]])
caso5= np.array([[2,1,6,5,22,2]])
casos=[caso1,caso2,caso3,caso4,caso5]

In [62]:
predicciones_arbol4= []
for i in casos:
    predicciones_arbol4.append(model.predict(i))

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 189ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 103ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 97ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 105ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 99ms/step


In [63]:
pred_array = np.vstack(predicciones_arbol4)   # queda shape (n, 22)

df_prediccion = pd.DataFrame(
    np.round(pred_array, 3),                 # Redondear a 3 decimales
    index = np.arange(1, pred_array.shape[0] + 1),   # Filas desde 1
    columns = np.arange(1, pred_array.shape[1] + 1)  # Columnas desde 1
)

df_prediccion

1      2
1  0.545  0.455
2  0.565  0.435
3  0.762  0.238
4  0.597  0.403
5  0.653  0.347

## Modelo 4: Predecir sexo

In [29]:
data_arbol5 = datos_consolidados[
    (datos_consolidados["fall_les"].isin([1, 2])) &
    (datos_consolidados["sexo_per"] != 9) &
    (datos_consolidados["tipo_eve"] != 99) &
       (datos_consolidados["tipo_veh"] != 99) &
    (datos_consolidados["edad_per"] != 999)].copy()

In [30]:
X = data_arbol5[['horario','mayor_menor','tipo_eve','tipo_veh','depto_ocu','fall_les']]
X = X.fillna(0)
y = data_arbol5['sexo_per'].values - 1

In [31]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

En este modelo se tienen 22 casos de salida y 5 variables de entrada

In [32]:
model = Sequential()
model.add(Dense(8, input_dim=6, activation='relu'))
model.add(Dense(4, activation='relu'))
model.add(Dense(2, activation='softmax'))

C:\Users\Brandon Portillo\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [33]:
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [34]:
model.fit(X_train, y_train, epochs=50, batch_size=64, validation_data=(X_test, y_test))

Epoch 1/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7108 - loss: 0.6037 - val_accuracy: 0.7195 - val_loss: 0.5873
Epoch 2/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7122 - loss: 0.5895 - val_accuracy: 0.7186 - val_loss: 0.5803
Epoch 3/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7120 - loss: 0.5814 - val_accuracy: 0.7207 - val_loss: 0.5720
Epoch 4/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7111 - loss: 0.5761 - val_accuracy: 0.7200 - val_loss: 0.5690
Epoch 5/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7106 - loss: 0.5743 - val_accuracy: 0.7191 - val_loss: 0.5680
Epoch 6/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7100 - loss: 0.5730 - val_accuracy: 0.7201 - val_loss: 0.5677
Epoch 7/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7116 - loss: 0.5723 - val_accuracy: 0.7192 - val_loss: 0.5665
Epoch 8/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7109 - loss: 0.5717 - val_accuracy: 0.

In [36]:
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Loss: {loss}, Accuracy: {accuracy}')

499/499 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7208 - loss: 0.5603
Loss: 0.5603396892547607, Accuracy: 0.720845639705658


El modelo acerto el 70% de las veces"

In [37]:
caso1= np.array([[1,1,2,4,1,1]])
caso2= np.array([[2,2,5,1,6,1]])
caso3= np.array([[1,2,3,3,14,2]])
caso4= np.array([[1,1,1,2,18,2]])
caso5= np.array([[2,1,6,5,22,2]])
casos=[caso1,caso2,caso3,caso4,caso5]

In [38]:
predicciones_arbol5= []
for i in casos:
    predicciones_arbol5.append(model.predict(i))


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 84ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 55ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 52ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 39ms/step


In [39]:
pred_array = np.vstack(predicciones_arbol5)   # queda shape (n, 22)

df_prediccion = pd.DataFrame(
    np.round(pred_array, 3),                 # Redondear a 3 decimales
    index = np.arange(1, pred_array.shape[0] + 1),   # Filas desde 1
    columns = np.arange(1, pred_array.shape[1] + 1)  # Columnas desde 1
)

df_prediccion


1      2
1  0.902  0.098
2  0.751  0.249
3  0.530  0.470
4  0.571  0.429
5  0.738  0.262

# Predicción de Depto

In [40]:
data_arbol5 = datos_consolidados[
    (datos_consolidados["fall_les"].isin([1, 2])) &
    (datos_consolidados["sexo_per"] != 9) &
    (datos_consolidados["tipo_eve"] != 99) &
       (datos_consolidados["tipo_veh"] != 99) &
       (datos_consolidados["depto_ocu"] != 1) &
    (datos_consolidados["edad_per"] != 999)].copy()

In [41]:
X = data_arbol5[['sexo_per','tipo_eve','tipo_veh','mayor_menor','fall_les']]
X = X.fillna(0)
y = data_arbol5['depto_ocu'].values - 1

In [42]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [43]:
model = Sequential()
model.add(Dense(16, input_dim=5, activation='relu'))
model.add(Dense(16, activation='relu'))
model.add(Dense(22, activation='softmax'))

C:\Users\Brandon Portillo\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [44]:
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [45]:
model.fit(X_train, y_train, epochs=50, batch_size=64, validation_data=(X_test, y_test))

Epoch 1/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.1108 - loss: 2.9946 - val_accuracy: 0.1316 - val_loss: 2.9497
Epoch 2/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.1305 - loss: 2.9434 - val_accuracy: 0.1305 - val_loss: 2.9408
Epoch 3/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.1308 - loss: 2.9348 - val_accuracy: 0.1294 - val_loss: 2.9338
Epoch 4/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.1313 - loss: 2.9296 - val_accuracy: 0.1299 - val_loss: 2.9323
Epoch 5/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.1312 - loss: 2.9263 - val_accuracy: 0.1318 - val_loss: 2.9284
Epoch 6/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.1329 - loss: 2.9233 - val_accuracy: 0.1328 - val_loss: 2.9283
Epoch 7/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.1339 - loss: 2.9208 - val_accuracy: 0.1368 - val_loss: 2.9236
Epoch 8/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.1354 - loss: 2.9188 - val_accuracy: 0.

In [46]:
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Loss: {loss}, Accuracy: {accuracy}')

329/329 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.1408 - loss: 2.9016
Loss: 2.9015636444091797, Accuracy: 0.14079147577285767


In [50]:
caso1= np.array([[2,2,4,1,1]])
caso2= np.array([[1,5,1,2,2]])
caso3= np.array([[2,3,3,2,2]])
caso4= np.array([[1,1,2,1,2]])
caso5= np.array([[2,6,5,1,2]])
casos=[caso1,caso2,caso3,caso4,caso5]

In [51]:
predicciones_arbol6= []
for i in casos:
    predicciones_arbol6.append(model.predict(i))


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 197ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 113ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 98ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 106ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 100ms/step


In [52]:
pred_array = np.vstack(predicciones_arbol6)   # queda shape (n, 22)

df_prediccion = pd.DataFrame(
    np.round(pred_array, 3),                 # Redondear a 3 decimales
    index = np.arange(1, pred_array.shape[0] + 1),   # Filas desde 1
    columns = np.arange(1, pred_array.shape[1] + 1)  # Columnas desde 1
)

df_prediccion

1      2      3      4      5      6      7      8      9      10  ...  \
1  0.0  0.033  0.044  0.052  0.159  0.059  0.019  0.009  0.040  0.039  ...   
2  0.0  0.037  0.062  0.079  0.123  0.059  0.051  0.025  0.063  0.033  ...   
3  0.0  0.042  0.026  0.055  0.055  0.050  0.038  0.020  0.045  0.042  ...   
4  0.0  0.054  0.050  0.055  0.119  0.061  0.031  0.022  0.066  0.042  ...   
5  0.0  0.016  0.058  0.049  0.130  0.031  0.027  0.021  0.043  0.037  ...   

      13     14     15     16     17     18     19     20     21     22  
1  0.034  0.029  0.035  0.081  0.087  0.074  0.036  0.037  0.027  0.048  
2  0.028  0.025  0.030  0.070  0.039  0.036  0.038  0.038  0.024  0.058  
3  0.067  0.094  0.042  0.105  0.054  0.040  0.041  0.041  0.029  0.045  
4  0.034  0.031  0.025  0.049  0.058  0.048  0.050  0.036  0.023  0.060  
5  0.030  0.032  0.038  0.165  0.033  0.070  0.035  0.018  0.058  0.045  

[5 rows x 22 columns]